compaction/compaction.cc

```diff
@@ -156,17 +172,31 @@ static api::timestamp_type get_max_purgeable_timestamp(const table_state& table_
         if (compacting_set.contains(sst)) {
             continue;
         }
+        api::timestamp_type min_timestamp = sst->get_stats_metadata().min_timestamp;
+        auto ts_stats = sst->get_ext_timestamp_stats();
+        if (!ts_stats.empty()) {
+            auto stat = is_shadowable ?
+                    sstables::ext_timestamp_stats_type::min_live_row_marker_timestamp :
+                    sstables::ext_timestamp_stats_type::min_live_timestamp;
+            auto it = ts_stats.find(stat);
+            if (it != ts_stats.end()) {
+                min_timestamp = it->second;
+            } else {
+                // Do not throw an expection in production, just use the legacy min_timestamp set above
+                on_internal_error_noexcept(clogger, format("Missing extended timestamp statstics: stat={} is_shadowable={}", int(stat), bool(is_shadowable)));
```

Comment:
> Please don't forgot that it is possible that both min_live_row_marker_timestamp and min_live_timestamp are missing, and this shouldn't be considered an on_internal_error. This can happen for example when the view had only deletions in this memtable but no new data.

Right, the internal error, is inside a `if (!ts_stats.empty)` block.

> I wonder maybe there might be a valid case of a view (where shadowable tombstone happen), which has a memtable that only modifies individual cells and not the row marker. Perhaps there can be, today or in the future, an optimization that doesn't modify the row marker if only a specific cell changed (see also https://github.com/scylladb/scylladb/issues/3594). So maybe your old fallback wasn't wrong? But to be honest, I'm not sure if this case can happen.

The problem is that the statistics cover all partitions in the memtable or sstable.
Initially, when I called this stat `shadowable_timestamp` this is roughly what I had in mind, i.e. any timestamp that is relevant for shadowable tombstones, but since we ended up tracking only live row_marker timestamps it made more sense to name it `min_live_row_marker_timestamp`.

Reply:
My comment is that you seem to assume that if both min_live_row_marker_timestamp and min_live_timestamp are missing (which is a **valid** case, and not reason for "internal error"), then you'll have ts_stats.empty(). I'm worried that perhaps in the future you'll need to add some third statistics for some other use case (maybe we'll invent some super-duper-tombstones and need some statistic of all even-numbered columns), and suddenly ts_stats.empty() will no longer be true. I think you should either defend against this case, or add a test for this case (e.g., involving only deletions with no new values) which will start to break if this problem ever happens (that ts_stats grew and is no longer empty).

Reply:
> My comment is that you seem to assume that if both min_live_row_marker_timestamp and min_live_timestamp are missing (which is a **valid** case, and not reason for "internal error"), then you'll have ts_stats.empty(). I'm worried that perhaps in the future you'll need to add some third statistics for some other use case (maybe we'll invent some super-duper-tombstones and need some statistic of all even-numbered columns), and suddenly ts_stats.empty() will no longer be true. I think you should either defend against this case, or add a test for this case (e.g., involving only deletions with no new values) which will start to break if this problem ever happens (that ts_stats grew and is no longer empty).

Can do.
In case of empty sstable, or one that has only deletes, ts_stats is expected to be non-empty and hold both min_live_timestamp and min_live_row_marker_timestamp. Both equal to api::max_timestamp.